File: scripts/lint_value_dashboard.py
#!/usr/bin/env python3
"""Lint `docs/value.md` for structural invariants.

Phase 5 Step 3 of `agents/roadmaps/road-to-readable-value-dashboard.md`.

Invariants enforced (any violation → exit 1):

1. Required sections present (intro / Reference scale / Panel A / Panel B
   / Glossar / NETTO line).
2. Every cost-ladder rung row cites a `source_report` (or `n/a` for the
   baseline rung) — no rung sneaks in without traceability.
3. No `measured` rung renders a `pending` source — internal consistency
   of confidence ↔ source state.
4. No negative-saving label: the literal string "Ersparnis" must not
   appear in a row where the displayed Δ-token value is positive (the
   load + terse rungs are *costs*, not savings; mislabelling either is
   a credibility failure the page explicitly forbids).
5. The `latest.json` exists and its `cost_ladder` rung ids match the
   five canonical rungs — the renderer cannot silently drop a rung.

The linter loads `internal/bench/reports/value/latest.json` directly
(not just the rendered `.md`) for items (3) and (5) — the rendered
text alone is too lossy.

Output: one violation per line in non-quiet mode; one-line summary in
quiet mode. Exit 0 on clean, 1 on any violation.
"""
from __future__ import annotations

import argparse
import json
import re
import sys
from pathlib import Path
from typing import Any, Dict, List


REPO_ROOT = Path(__file__).resolve().parent.parent
DASHBOARD = REPO_ROOT / "docs" / "value.md"
LATEST = REPO_ROOT / "internal" / "bench" / "reports" / "value" / "latest.json"

REQUIRED_SECTIONS = (
    "# Value Dashboard",
    "## Reference scale",
    "## Panel A",
    "## Panel B",
    "## Glossar",
    "**NETTO",
)

CANONICAL_RUNG_IDS = ("baseline", "load", "thin", "condense", "rtk", "terse")


def _log(msg: str, quiet: bool, *, err: bool = False) -> None:
    if err:
        print(msg, file=sys.stderr)
    elif not quiet:
        print(msg)


def check_required_sections(text: str) -> List[str]:
    return [
        f"missing required section: '{section}'"
        for section in REQUIRED_SECTIONS
        if section not in text
    ]


def check_source_citations(report: Dict[str, Any]) -> List[str]:
    violations = []
    for rung in report.get("cost_ladder", []) or []:
        source = rung.get("source_report")
        if not source:
            violations.append(
                f"rung '{rung.get('id')}' has no source_report field"
            )
            continue
        if not isinstance(source, str) or not source.strip():
            violations.append(
                f"rung '{rung.get('id')}' has empty source_report"
            )
    return violations


def check_confidence_vs_source(report: Dict[str, Any]) -> List[str]:
    """A `measured` rung's source_report must exist on disk."""
    violations = []
    for rung in report.get("cost_ladder", []) or []:
        if rung.get("confidence") != "measured":
            continue
        source = rung.get("source_report") or ""
        if source in ("", "n/a"):
            continue  # baseline rung
        path = REPO_ROOT / source
        if not path.exists():
            violations.append(
                f"rung '{rung.get('id')}' is 'measured' but its "
                f"source_report does not exist: {source}"
            )
    return violations


def check_no_negative_savings(text: str) -> List[str]:
    """A rung whose Δ-token value is positive must not be labelled a saving.

    Heuristic: scan Panel A's rows; flag any row that contains the
    German word "Ersparnis" with a positive token-delta in the same row.
    """
    violations = []
    # Panel A rows are pipe-delimited; we read every line starting with "|"
    # inside the cost ladder section.
    in_panel_a = False
    for line in text.splitlines():
        if line.startswith("## Panel A"):
            in_panel_a = True
            continue
        if in_panel_a and line.startswith("## "):
            break
        if not in_panel_a or not line.startswith("|"):
            continue
        if "Ersparnis" not in line:
            continue
        # Look for a "+" sign at the start of an integer-shaped delta.
        # The format renders deltas as "+4 843" / "-186".
        m = re.search(r"\|\s*([+-][0-9 ]+)\s*\|", line)
        if m and m.group(1).strip().startswith("+"):
            token_value = m.group(1).strip()
            violations.append(
                "row labelled 'Ersparnis' has a positive Δ-token value: "
                f"{token_value!r} — positive deltas are costs, not savings."
            )
    return violations


def check_canonical_rung_set(report: Dict[str, Any]) -> List[str]:
    rungs = report.get("cost_ladder", []) or []
    ids = [r.get("id") for r in rungs]
    if list(ids) != list(CANONICAL_RUNG_IDS):
        return [
            f"cost_ladder rung ids must be {CANONICAL_RUNG_IDS}, "
            f"got {tuple(ids)}"
        ]
    return []


def lint(quiet: bool = False) -> int:
    violations: List[str] = []

    if not DASHBOARD.exists():
        _log(
            f"FAIL: dashboard not found: {DASHBOARD.relative_to(REPO_ROOT)}",
            quiet,
            err=True,
        )
        return 1
    text = DASHBOARD.read_text()
    violations.extend(check_required_sections(text))
    violations.extend(check_no_negative_savings(text))

    if not LATEST.exists():
        # No JSON to deep-check — that's a placeholder dashboard.
        # Required-sections check still applies; we degrade gracefully.
        if violations:
            for v in violations:
                _log(f"FAIL: {v}", quiet, err=True)
            return 1
        _log(
            "lint_value_dashboard: dashboard is a placeholder "
            "(no value-v1.json yet) — structural checks pass.",
            quiet=False,
        )
        return 0

    try:
        report = json.loads(LATEST.read_text())
    except json.JSONDecodeError as exc:
        _log(f"FAIL: {LATEST.name} is not valid JSON: {exc}", quiet, err=True)
        return 1

    violations.extend(check_source_citations(report))
    violations.extend(check_confidence_vs_source(report))
    violations.extend(check_canonical_rung_set(report))

    if violations:
        for v in violations:
            _log(f"FAIL: {v}", quiet, err=True)
        return 1
    _log(
        (
            "lint_value_dashboard: OK — "
            f"{len(report.get('cost_ladder', []))} rungs, "
            f"{len(report.get('behaviour', []))} behaviour metrics, all "
            "sections present, all sources cited."
        ),
        quiet=False,
    )
    return 0


def parse_args(argv: List[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Lint docs/value.md for structural invariants."
    )
    parser.add_argument(
        "--quiet",
        action="store_true",
        help="Suppress non-error output.",
    )
    return parser.parse_args(argv)


def main(argv: List[str] | None = None) -> int:
    args = parse_args(argv if argv is not None else sys.argv[1:])
    return lint(quiet=args.quiet)


if __name__ == "__main__":
    raise SystemExit(main())

File: src/scripts/lint_value_dashboard.ts
#!/usr/bin/env tsx
/**
 * Lint `docs/value.md` for structural invariants.
 *
 * TypeScript twin of `src/scripts/lint_value_dashboard.py` (ADR-096,
 * Phase 4 / Wave 4b). The CLI contract is mirrored EXACTLY — `--quiet`
 * flag, exit codes (0 / 1), stdout/stderr split, byte-identical messages,
 * same check order. No behaviour changes — latent bugs replicated.
 *
 * Output: one violation per line in non-quiet mode; one-line summary in
 * quiet mode. Exit 0 on clean, 1 on any violation.
 *
 * DIVERGENCE CANDIDATE: the malformed-JSON failure path prints the host
 * JSON parser's exception text (`FAIL: latest.json is not valid JSON: …`),
 * which differs between Python's `json.JSONDecodeError` and JS `JSON.parse`.
 * The happy path and every structural check are byte-identical.
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';

const _HERE = fileURLToPath(import.meta.url);
const REPO_ROOT = path.resolve(path.dirname(_HERE), '..', '..');
const DASHBOARD = path.join(REPO_ROOT, 'docs', 'value.md');
const LATEST = path.join(
    REPO_ROOT,
    'internal',
    'bench',
    'reports',
    'value',
    'latest.json',
);

const REQUIRED_SECTIONS = [
    '# Value Dashboard',
    '## Reference scale',
    '## Panel A',
    '## Panel B',
    '## Glossary',
    '**NET',
] as const;

const CANONICAL_RUNG_IDS = ['baseline', 'load', 'thin', 'condense', 'rtk', 'terse'] as const;

type JsonObject = Record<string, unknown>;

function _exists(p: string): boolean {
    try {
        fs.statSync(p);
        return true;
    } catch {
        return false;
    }
}

function _relPosix(target: string, root: string): string {
    return path.relative(root, target).split(path.sep).join('/');
}

function _log(msg: string, quiet: boolean, err = false): void {
    if (err) {
        process.stderr.write(msg + '\n');
    } else if (!quiet) {
        process.stdout.write(msg + '\n');
    }
}

function check_required_sections(text: string): string[] {
    return REQUIRED_SECTIONS.filter((section) => !text.includes(section)).map(
        (section) => `missing required section: '${section}'`,
    );
}

function _asArray(v: unknown): unknown[] {
    // Mirror `report.get("cost_ladder", []) or []` — a falsy / absent value
    // becomes an empty list; a list passes through.
    if (Array.isArray(v)) {
        return v;
    }
    return [];
}

function check_source_citations(report: JsonObject): string[] {
    const violations: string[] = [];
    for (const rungU of _asArray(report['cost_ladder'])) {
        const rung = rungU as JsonObject;
        const source = rung['source_report'];
        if (!source) {
            violations.push(`rung '${_g(rung['id'])}' has no source_report field`);
            continue;
        }
        if (typeof source !== 'string' || !source.trim()) {
            violations.push(`rung '${_g(rung['id'])}' has empty source_report`);
        }
    }
    return violations;
}

/** A `measured` rung's source_report must exist on disk. */
function check_confidence_vs_source(report: JsonObject): string[] {
    const violations: string[] = [];
    for (const rungU of _asArray(report['cost_ladder'])) {
        const rung = rungU as JsonObject;
        if (rung['confidence'] !== 'measured') {
            continue;
        }
        const source = (rung['source_report'] as string) || '';
        if (source === '' || source === 'n/a') {
            continue; // baseline rung
        }
        const p = path.join(REPO_ROOT, source);
        if (!_exists(p)) {
            violations.push(
                `rung '${_g(rung['id'])}' is 'measured' but its ` +
                    `source_report does not exist: ${source}`,
            );
        }
    }
    return violations;
}

function check_no_negative_savings(text: string): string[] {
    const violations: string[] = [];
    let in_panel_a = false;
    for (const line of text.split('\n')) {
        if (line.startsWith('## Panel A')) {
            in_panel_a = true;
            continue;
        }
        if (in_panel_a && line.startsWith('## ')) {
            break;
        }
        if (!in_panel_a || !line.startsWith('|')) {
            continue;
        }
        if (!line.includes('Ersparnis')) {
            continue;
        }
        const m = /\|\s*([+-][0-9 ]+)\s*\|/.exec(line);
        if (m && m[1]!.trim().startsWith('+')) {
            const token_value = m[1]!.trim();
            violations.push(
                "row labelled 'Ersparnis' has a positive Δ-token value: " +
                    `${_pyRepr(token_value)} — positive deltas are costs, not savings.`,
            );
        }
    }
    return violations;
}

function check_canonical_rung_set(report: JsonObject): string[] {
    const rungs = _asArray(report['cost_ladder']);
    const ids = rungs.map((r) => (r as JsonObject)['id']);
    if (!_seqEqual(ids, CANONICAL_RUNG_IDS as readonly unknown[])) {
        return [
            `cost_ladder rung ids must be ${_pyTupleRepr(CANONICAL_RUNG_IDS)}, ` +
                `got ${_pyTupleRepr(ids)}`,
        ];
    }
    return [];
}

function _seqEqual(a: readonly unknown[], b: readonly unknown[]): boolean {
    if (a.length !== b.length) {
        return false;
    }
    for (let i = 0; i < a.length; i++) {
        if (a[i] !== b[i]) {
            return false;
        }
    }
    return true;
}

function lint(quiet = false): number {
    const violations: string[] = [];

    if (!_exists(DASHBOARD)) {
        _log(`FAIL: dashboard not found: ${_relPosix(DASHBOARD, REPO_ROOT)}`, quiet, true);
        return 1;
    }
    const text = fs.readFileSync(DASHBOARD, 'utf-8');
    violations.push(...check_required_sections(text));
    violations.push(...check_no_negative_savings(text));

    if (!_exists(LATEST)) {
        if (violations.length) {
            for (const v of violations) {
                _log(`FAIL: ${v}`, quiet, true);
            }
            return 1;
        }
        _log(
            'lint_value_dashboard: dashboard is a placeholder ' +
                '(no value-v1.json yet) — structural checks pass.',
            false,
        );
        return 0;
    }

    let report: JsonObject;
    try {
        report = JSON.parse(fs.readFileSync(LATEST, 'utf-8')) as JsonObject;
    } catch (exc) {
        const msg = exc instanceof Error ? exc.message : String(exc);
        _log(`FAIL: ${path.basename(LATEST)} is not valid JSON: ${msg}`, quiet, true);
        return 1;
    }

    violations.push(...check_source_citations(report));
    violations.push(...check_confidence_vs_source(report));
    violations.push(...check_canonical_rung_set(report));

    if (violations.length) {
        for (const v of violations) {
            _log(`FAIL: ${v}`, quiet, true);
        }
        return 1;
    }
    _log(
        'lint_value_dashboard: OK — ' +
            `${_asArray(report['cost_ladder']).length} rungs, ` +
            `${_asArray(report['behaviour']).length} behaviour metrics, all ` +
            'sections present, all sources cited.',
        false,
    );
    return 0;
}

/** Mirror Python's f-string rendering of a value used in `rung.get('id')`. */
function _g(v: unknown): string {
    if (v === undefined || v === null) {
        return 'None';
    }
    return String(v);
}

/** Mirror Python repr() of a string. */
function _pyRepr(s: string): string {
    const hasSingle = s.includes("'");
    const hasDouble = s.includes('"');
    const quote = hasSingle && !hasDouble ? '"' : "'";
    let body = s.replace(/\\/g, '\\\\');
    if (quote === "'") {
        body = body.replace(/'/g, "\\'");
    } else {
        body = body.replace(/"/g, '\\"');
    }
    return `${quote}${body}${quote}`;
}

/** Mirror Python repr() of a tuple of (possibly non-string) values. */
function _pyTupleRepr(items: readonly unknown[]): string {
    const parts = items.map((i) => {
        if (typeof i === 'string') {
            return _pyRepr(i);
        }
        if (i === undefined || i === null) {
            return 'None';
        }
        return String(i);
    });
    if (parts.length === 1) {
        return `(${parts[0]},)`;
    }
    return `(${parts.join(', ')})`;
}

function parse_args(argv: readonly string[]): boolean {
    let quiet = false;
    for (const arg of argv) {
        if (arg === '--quiet') {
            quiet = true;
        } else if (arg === '-h' || arg === '--help') {
            process.stdout.write('usage: lint_value_dashboard.py [-h] [--quiet]\n');
            process.exit(0);
        } else {
            process.stderr.write(
                `usage: lint_value_dashboard.py [-h] [--quiet]\nlint_value_dashboard.py: error: unrecognized arguments: ${arg}\n`,
            );
            process.exit(2);
        }
    }
    return quiet;
}

function main(argv?: readonly string[]): number {
    const args = parse_args(argv ?? process.argv.slice(2));
    return lint(args);
}

const _isCliEntry =
    process.argv[1] !== undefined &&
    import.meta.url === pathToFileURL(path.resolve(process.argv[1])).href;
if (_isCliEntry || process.argv[1] === _HERE) {
    process.exit(main());
}

export {
    type JsonObject,
    REPO_ROOT,
    DASHBOARD,
    LATEST,
    REQUIRED_SECTIONS,
    CANONICAL_RUNG_IDS,
    check_required_sections,
    check_source_citations,
    check_confidence_vs_source,
    check_no_negative_savings,
    check_canonical_rung_set,
    lint,
    main,
};
